docs/design/import.md

```diff
@@ -0,0 +1,314 @@
+
+    // StateContext is called to convert an ID to one or more InstanceState to
+    // insert into the Terraform state. If this isn't specified, then
+    // the ID is passed straight through. This function receives a context
+    // that will cancel if Terraform sends a cancellation signal.
+    StateContext StateContextFunc
+}
+
+// StateFunc is the function called to import a resource into the Terraform state.
+//
+// Deprecated: Please use the context aware equivalent StateContextFunc.
+type StateFunc func(*ResourceData, interface{}) ([]*ResourceData, error)
+
+// StateContextFunc is the function called to import a resource into the
+// Terraform state. It is given a ResourceData with only ID set. This
+// ID is going to be an arbitrary value given by the user and may not map
+// directly to the ID format that the resource expects, so that should
+// be validated.
+//
+// This should return a slice of ResourceData that turn into the state
+// that was imported. This might be as simple as returning only the argument
+// that was given to the function. In other cases (such as AWS security groups),
+// an import may fan out to multiple resources and this will have to return
+// multiple.
+//
+// To create the ResourceData structures for other resource types (if
+// you have to), instantiate your resource and call the Data function.
+type StateContextFunc func(context.Context, *ResourceData, interface{}) ([]*ResourceData, error)
+```
+
+When defined, the resource could respond to the `ImportResource` RPC, otherwise an error was returned that the resource does not support import.
+
+If the resource state could be entirely fetched using an import identifier that matched the resource identifier, the `ImportStatePassthrough` and `ImportStatePassthroughContext` helpers simplified provider implementations, e.g.
+
+```go
+Importer: &schema.ResourceImporter{
+    State: schema.ImportStatePassthrough,
+},
+```
+
+Otherwise, custom provider implementations were required.
+
+## Caveats
+
+Terraform supports the ability to import multiple resource states during a single import (for example, resource import of an `aws_s3_bucket` also imports any associated `aws_s3_bucket_policy`), however almost all resources in the Terraform ecosystem with import support implement a 1:1 import. The extremely uncommon multiple resource support in practice is likely due to:
+
+- Risk of resource destruction for incorrect implementations ([example issue reference](https://github.com/hashicorp/terraform-provider-aws/issues/6036)).
+- Pracitioners may not have required permissions or wish to manage secondary imported resources ([example issue reference](https://github.com/hashicorp/terraform-provider-aws/issues/9508)).
+- Providers must define the resource addresses for secondary resources, which may not match practitioner desires and require additional action such as `state mv` ([example issue reference](https://github.com/hashicorp/terraform-provider-aws/issues/9001)).
```

Comment:
This feels like an important thing to call out. How does this work currently? It sounds like the provider can supply the resource address, but I'm not clear on how. Can we tease out this bit a little more to be more concrete about what the state of the world is right now?

Reply:
The latest commit now includes lots more information about multiple resource import including what Terraform CLI does implicitly with resource addresses and some examples of how support was implemented in the previous framework. 👍